chapters/part3/document-ranking.md

```diff
@@ -348,7 +348,91 @@ Next, we need to evaluate the model's performance. There are several metrics com
 | elmo           | 0.4 (2)           | 0.55 (1)               |
 | grover         | 0.1 (3)           | 0.05 (3)               |
 
-The correct option is the cookie monster, but our model only ranked it second. It shouldn't get a complete zero for that, but shouldn't be fully rewarded either. The reciprocal rank would give this a `1/2` for ranking the best option 2nd. More generally, the reciprocal rank score is `1` divided by the position your model ranked the correct option. If it had correctly ranked cookie monster first, its score would be `1/1 = 1`. As the name suggests, we take the mean of all these reciprocal rank scores across our dataset to find the MRR.
+The correct option is the cookie monster, but our model only ranked it second. It shouldn't get a complete zero for that, but shouldn't be fully rewarded either. The reciprocal rank would give this a `1/2` for ranking the best option 2nd. More generally, the reciprocal rank score is `1` divided by the position your model ranked the correct option. If it had correctly ranked cookie monster first, its score would be `1/1 = 1`. As the name suggests, we take the mean of all these reciprocal rank scores across our dataset to find the MRR. This isn't too hard to implement in Pytorch:
 
+```python
+def first_nonzero(t):
+    idx = torch.arange(t.size(1), 0, -1).type_as(t)
+    indices = torch.argmax(t * idx, 1, keepdim=True)
+    return indices
+
+def mrr(y_pred, y_true, mask):
+    y_pred = y_pred.masked_fill(~mask, -1)
+    y_true = y_true.ge(y_true.max(dim=-1, keepdim=True).values).long()
```

Comment:
This line is pretty cryptic to me.  In general, I spent a whole lot of time reading through this logic trying to decide what exactly it was doing.  I eventually figured it out, but some comments describing the logic would be very helpful to the reader.

Reply:
Added a lot of comments here, let me know if this helps. To be honest, this is a bit of a difficult metric to compute with matrix operations while also keeping the code clear and easily understandable (I feel like that applies to most ranking metrics).